lib/connection/connection.js

```diff
@@ -528,97 +528,142 @@ function merge(options1, options2) {
-    // Error on auth or skip
-    if (connection.authorizationError && self.rejectUnauthorized) {
-      return self.emit('error', connection.authorizationError, self, { ssl: true });
+    // Set default sni servername to be the same as host
+    if (options.servername == null) {
+      options.servername = self.host;
     }
 
+    options = Object.assign({}, options, { host: self.host, port: self.port });
+
+    return options;
+  } else {
+    if (self.domainSocket) {
+      options = { path: self.host };
+    } else {
+      options = { port: self.port, host: self.host };
+    }
+    return options;
+  }
+}
+
+function makeConnection(self, options, callback) {
+  const module = options.ssl ? tls : net;
+
+  const connection = module.connect(options, function() {
+    if (self.ssl) {
+      // Error on auth or skip
+      if (connection.authorizationError && self.rejectUnauthorized) {
+        return self.emit('error', connection.authorizationError, self, { ssl: true });
+      }
+    }
     // Set socket timeout instead of connection timeout
+
     connection.setTimeout(self.socketTimeout);
-    // We are done emit connect
-    self.emit('connect', self);
+    return callback(null, connection);
   });
 
   // Set the options for the connection
   connection.setKeepAlive(self.keepAlive, self.keepAliveInitialDelay);
   connection.setTimeout(self.connectionTimeout);
   connection.setNoDelay(self.noDelay);
 
-  return connection;
+  // Add handlers for events
+  connection.once('error', err =>
+    callback({ err: err, type: 'error', family: options.family }, null)
+  );
+  connection.once('timeout', err => callback({ err: err, type: 'timeout' }, null));
+  connection.once('close', err => callback({ err: err, type: 'close' }, null));
+  connection.on('data', err => callback({ err: err, type: 'data' }, null));
+  return;
 }
 
-function makeUnsecureConnection(self, family) {
-  // Create new connection instance
-  let connection_options;
-  if (self.domainSocket) {
-    connection_options = { path: self.host };
-  } else {
-    connection_options = { port: self.port, host: self.host };
-    connection_options.family = family;
-  }
-
-  const connection = net.createConnection(connection_options);
-
-  // Set the options for the connection
-  connection.setKeepAlive(self.keepAlive, self.keepAliveInitialDelay);
-  connection.setTimeout(self.connectionTimeout);
-  connection.setNoDelay(self.noDelay);
-
-  connection.once('connect', function() {
-    // Set socket timeout instead of connection timeout
-    connection.setTimeout(self.socketTimeout);
-    // Emit connect event
-    self.emit('connect', self);
-  });
+function fastFallback(self, _options, callback) {
+  const options = prepareConnectionOptions(self, _options);
+
+  let connection;
+  const connectionHandler = (err, _connection) => {
+    const _errorHandler = errorHandler(self);
+    const _timeoutHandler = timeoutHandler(self);
```

Comment:
I don't think you need to create these handlers here, they will be created later when you attach them to the returned connection. Here we are _only_ trying to make the connections, focusing on listening to them properly later.